File: demo-deno/xlsx2json/nodejs/batch-convert.js
#!/usr/bin/env node

/**
 * XLSX转JSON批量处理脚本 - Node.js版本
 * 用法: node batch-convert.js [目录路径]
 *
 * 此脚本会：
 * 1. 查找指定目录（默认当前目录）中的所有.xlsx文件
 * 2. 显示转换选项给用户选择
 * 3. 执行相应的转换操作
 */

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function findXlsxFiles(directory = '.') {
    const files = [];

    try {
        const entries = fs.readdirSync(directory, { withFileTypes: true });

        for (const entry of entries) {
            if (entry.isFile() && entry.name.toLowerCase().endsWith('.xlsx')) {
                files.push(entry.name);
            }
        }
    } catch (error) {
        throw new Error(`无法读取目录 ${directory}: ${error.message}`);
    }

    return files;
}

function showFiles(files) {
    console.log(`找到 ${files.length} 个Excel文件:`);
    files.forEach((file, index) => {
        console.log(`  ${index + 1}. ${file}`);
    });
}

function generateOutputName(inputFile, extension) {
    const parsed = path.parse(inputFile);
    return parsed.name + extension;
}

function showConvertOptions() {
    console.log("\n转换选项:");
    console.log("1. 使用完整版本（需要安装xlsx依赖包）");
    console.log("2. 手动转换指南（离线方式）");
    console.log("3. 退出");
}

function promptUser(question) {
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            resolve(answer.trim());
        });
    });
}

function runCommand(command, args) {
    return new Promise((resolve, reject) => {
        const process = spawn(command, args, { stdio: 'inherit' });

        process.on('close', (code) => {
            if (code === 0) {
                resolve();
            } else {
                reject(new Error(`命令执行失败，退出码: ${code}`));
            }
        });

        process.on('error', (error) => {
            reject(error);
        });
    });
}

async function convertWithNodejs(files) {
    console.log("\n使用Node.js版本转换...");

    // 检查是否安装了依赖
    if (!fs.existsSync(path.join(__dirname, 'node_modules'))) {
        console.log("需要安装依赖包，正在执行 npm install...");
        try {
            await runCommand('npm', ['install']);
            console.log("✅ 依赖安装完成");
        } catch (error) {
            console.error("❌ 依赖安装失败:", error.message);
            console.log("请手动运行: npm install");
            return;
        }
    }

    for (const file of files) {
        try {
            console.log(`\n正在转换: ${file}`);
            await runCommand('node', ['xlsx2json.js', file]);
            console.log("✅ 转换成功");

        } catch (error) {
            console.error(`❌ 转换 ${file} 时出错: ${error.message}`);
        }
    }
}

function showManualGuide(files) {
    console.log("\n📋 手动转换指南:");
    console.log("====================");

    files.forEach((file, index) => {
        console.log(`\n${index + 1}. 处理文件: ${file}`);
        console.log("   步骤:");
        console.log("   a) 用Excel打开此文件");
        console.log("   b) 确保A列是key，B列是value");
        console.log("   c) 选择 文件 > 另存为");
        console.log(`   d) 文件名: ${generateOutputName(file, '.csv')}`);
        console.log("   e) 文件类型选择: CSV (逗号分隔)(*.csv)");
        console.log("   f) 点击保存");
        console.log(`   g) 运行: node xlsx2json-simple.js "${generateOutputName(file, '.csv')}"`);
    });

    console.log("\n或者，您也可以使用在线工具:");
    console.log("- https://convertio.co/xlsx-csv/");
    console.log("- https://www.zamzar.com/convert/xlsx-to-csv/");

    console.log("\n转换为CSV后，使用以下命令:");
    files.forEach(file => {
        const csvName = generateOutputName(file, '.csv');
        console.log(`node xlsx2json-simple.js "${csvName}"`);
    });
}

async function main() {
    const args = process.argv.slice(2);
    const directory = args[0] || '.';

    console.log("XLSX转JSON批量处理工具 - Node.js版本");
    console.log("===================================");

    try {
        // 查找Excel文件
        const files = await findXlsxFiles(directory);

        if (files.length === 0) {
            console.log(`在目录 "${directory}" 中没有找到.xlsx文件`);
            rl.close();
            return;
        }

        showFiles(files);
        showConvertOptions();

        const choice = await promptUser("\n请选择操作 (1-3): ");

        switch (choice) {
            case '1':
                await convertWithNodejs(files);
                break;
            case '2':
                showManualGuide(files);
                break;
            case '3':
                console.log("退出程序");
                break;
            default:
                console.log("无效选择");
                break;
        }

        rl.close();

    } catch (error) {
        console.error(`❌ ${error.message}`);
        rl.close();
        process.exit(1);
    }
}

// 只有在直接运行此脚本时才执行main函数
if (require.main === module) {
    main();
}

File: demo-deno/xlsx2json/python/batch_convert.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
XLSX转JSON批量处理脚本 - Python版本
用法: python batch_convert.py [目录路径]

此脚本会：
1. 查找指定目录（默认当前目录）中的所有.xlsx文件
2. 显示转换选项给用户选择
3. 执行相应的转换操作
"""

import sys
import os
import subprocess
from pathlib import Path


def find_xlsx_files(directory="."):
    """查找Excel文件"""
    files = []

    try:
        path = Path(directory)
        for file_path in path.iterdir():
            if file_path.is_file() and file_path.suffix.lower() == '.xlsx':
                files.append(file_path.name)
    except Exception as error:
        raise Exception(f"无法读取目录 {directory}: {error}")

    return files


def show_files(files):
    """显示找到的文件"""
    print(f"找到 {len(files)} 个Excel文件:")
    for i, file in enumerate(files, 1):
        print(f"  {i}. {file}")


def generate_output_name(input_file, extension):
    """生成输出文件名"""
    path = Path(input_file)
    return path.stem + extension


def show_convert_options():
    """显示转换选项"""
    print("\n转换选项:")
    print("1. 使用完整版本（需要安装openpyxl依赖包）")
    print("2. 手动转换指南（离线方式）")
    print("3. 退出")


def prompt_user(question):
    """提示用户输入"""
    try:
        return input(question).strip()
    except KeyboardInterrupt:
        print("\n\n程序被用户中断")
        sys.exit(0)


def run_command(command):
    """运行命令"""
    try:
        result = subprocess.run(command, shell=True, check=True,
                               capture_output=True, text=True)
        return True, result.stdout
    except subprocess.CalledProcessError as e:
        return False, e.stderr


def check_dependencies():
    """检查依赖"""
    try:
        import openpyxl
        return True
    except ImportError:
        return False


def install_dependencies():
    """安装依赖"""
    print("需要安装依赖包，正在执行 pip install openpyxl...")
    success, output = run_command("pip install openpyxl")

    if success:
        print("✅ 依赖安装完成")
        return True
    else:
        print("❌ 依赖安装失败:")
        print(output)
        print("请手动运行: pip install openpyxl")
        return False


def convert_with_python(files):
    """使用Python版本转换"""
    print("\n使用Python版本转换...")

    # 检查依赖
    if not check_dependencies():
        if not install_dependencies():
            return

    for file in files:
        try:
            print(f"\n正在转换: {file}")

            command = f'python xlsx2json.py "{file}"'
            success, output = run_command(command)

            if success:
                print("✅ 转换成功")
                print(output)
            else:
                print("❌ 转换失败:")
                print(output)

        except Exception as error:
            print(f"❌ 转换 {file} 时出错: {error}")


def show_manual_guide(files):
    """显示手动转换指南"""
    print("\n📋 手动转换指南:")
    print("====================")

    for i, file in enumerate(files, 1):
        print(f"\n{i}. 处理文件: {file}")
        print("   步骤:")
        print("   a) 用Excel打开此文件")
        print("   b) 确保A列是key，B列是value")
        print("   c) 选择 文件 > 另存为")
        print(f"   d) 文件名: {generate_output_name(file, '.csv')}")
        print("   e) 文件类型选择: CSV (逗号分隔)(*.csv)")
        print("   f) 点击保存")
        print(f"   g) 运行: python xlsx2json_simple.py \"{generate_output_name(file, '.csv')}\"")

    print("\n或者，您也可以使用在线工具:")
    print("- https://convertio.co/xlsx-csv/")
    print("- https://www.zamzar.com/convert/xlsx-to-csv/")

    print("\n转换为CSV后，使用以下命令:")
    for file in files:
        csv_name = generate_output_name(file, '.csv')
        print(f"python xlsx2json_simple.py \"{csv_name}\"")


def main():
    """主函数"""
    args = sys.argv[1:]
    directory = args[0] if args else "."

    print("XLSX转JSON批量处理工具 - Python版本")
    print("==================================")

    try:
        # 查找Excel文件
        files = find_xlsx_files(directory)

        if not files:
            print(f"在目录 \"{directory}\" 中没有找到.xlsx文件")
            return

        show_files(files)
        show_convert_options()

        choice = prompt_user("\n请选择操作 (1-3): ")

        if choice == '1':
            convert_with_python(files)
        elif choice == '2':
            show_manual_guide(files)
        elif choice == '3':
            print("退出程序")
        else:
            print("无效选择")

    except Exception as error:
        print(f"❌ {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()